ext/mbstring/mbstring.c

```diff
@@ -2443,12 +2443,21 @@ PHP_FUNCTION(mb_strcut)
 
 	if (enc->cut) {
 		RETURN_STR(enc->cut(string.val, from, len, string.val + string.len));
-	} else {
-		ret = mbfl_strcut(&string, &result, from, len);
-		ZEND_ASSERT(ret != NULL);
-		RETVAL_STRINGL((char *)ret->val, ret->len); /* the string is already strdup()'ed */
-		efree(ret->val);
 	}
+
+	unsigned int char_len = string.encoding->flag & (MBFL_ENCTYPE_SBCS | MBFL_ENCTYPE_WCS2 | MBFL_ENCTYPE_WCS4);
+	if (char_len) {
+		from &= -char_len;
+		if (len > string.len - from) {
+			len = string.len - from;
+		}
+		RETURN_STR(zend_string_init_fast((const char*)(string.val + from), len & -char_len));
```

Comment:
I'm slightly confused as to what the ``len & -char_len`` does

Reply:
Note that numbers are stored as two-complement. So that means -x == ~(x-1).
`char_len` is either 1, 2 or 4. so `-char_len` is either 0b111...11 (for 1), 0b111...10 (for 2), or 0b111...00 (for 4).
So when you do an AND, you either keep all bits (for 1), remove the last bit (for 2), or remove the two last bits (for 4).
IOW, this "rounds down" to the nearest multiple of 1, 2 or 4.

Reply:
> Note that numbers are stored as two-complement. So that means -x == ~(x-1). `char_len` is either 1, 2 or 4. so `-char_len` is either 0b111...11 (for 1), 0b111...10 (for 2), or 0b111...00 (for 4). So when you do an AND, you either keep all bits (for 1), remove the last bit (for 2), or remove the two last bits (for 4). IOW, this "rounds down" to the nearest multiple of 1, 2 or 4.

Ah this makes sense, but having this as a comment would be great for the less bit inclined people (like me) that might need to read this in the future :)